### PR TITLE
add guest caching to erdnest

### PR DESCRIPTION
**Pull request by florincoz**

### CODE REVIEW

File: packages/common/src/common/caching/guest-caching/guest-caching.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    res.setHeader('X-Guest-Cache-Hit', !!cacheResponse);
+
+    // Delete data for previous minute
+    if (cacheHitsCounter[previousMinute]) {
+      delete cacheHitsCounter[previousMinute];
+    }
+
+    if (cacheResponse) {
+      return res.json(cacheResponse);
+    } else {
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.middleware.ts:91]: remove else

File: packages/common/src/common/caching/guest-caching/guest-caching.service.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  public async recompute(options: IGuestCacheOptions) {
+    // recompute cache
+    const currentDate = moment().format(DATE_FORMAT);
+    const previousMinute = moment().subtract(1, 'minute').format(DATE_FORMAT);
+    const threshold = Number(options.cacheTriggerHitsThreshold || 100);
+    const keysToComputeCurrentMinute: string[] = await this.cachingService.zRange(`${REDIS_PREFIX}.${currentDate}`, threshold, '+inf', 'BYSCORE');
+    const keysToComputePreviousMinute: string[] = await this.cachingService.zRange(`${REDIS_PREFIX}.${previousMinute}`, threshold, '+inf', 'BYSCORE');
+
+    const keysToCompute = [...new Set([...keysToComputeCurrentMinute, ...keysToComputePreviousMinute])];
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.service.ts:42]: use `.distinct()` from array extensions

File: packages/common/src/common/caching/guest-caching/guest-caching.service.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    await Promise.allSettled(keysToCompute.map(async key => {
+      const parsedKey = `${REDIS_PREFIX}.${key}.body`;
+      const keyValue: IGuestCacheEntity | undefined = await this.cachingService.getCache(parsedKey);
+
+      if (!keyValue) {
+        return Promise.resolve();
+      }
+
+      console.log(`Started warming up query '${JSON.stringify(keyValue)}' for url '${options.targetUrl}'`);
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.service.ts:52]: use `Logger`

File: packages/common/src/common/caching/guest-caching/guest-caching.service.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      try {
+        const url = `${options.targetUrl}${keyValue.path}`;
+
+        if (keyValue.method === GuestCacheMethodEnum.GET) {
+          data = await this.getReq(url);
+        } else {
+          data = await this.postReq(url, keyValue.body);
+        }
+      } catch (error) {
+        console.error(`An error occurred while warming up query '${JSON.stringify(keyValue)}' for url '${options.targetUrl}'`);
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.service.ts:65]: use `Logger`

File: packages/common/src/common/caching/guest-caching/guest-caching.service.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        console.error(error);
+      }
+
+      profiler.stop();
+
+      console.log(`Finished warming up query '${JSON.stringify(keyValue)}' for url '${options.targetUrl}'. Response size: ${JSON.stringify(data).length}. Duration: ${profiler.duration}`);
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.service.ts:71]: use `Logger`

File: packages/common/src/common/caching/guest-caching/guest-caching.middleware.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    MetricsService.incrementGuestHits();
+
+    const currentMinute = moment().format(DATE_FORMAT);
+    const previousMinute = moment().subtract(1, 'minute').format(DATE_FORMAT);
+    const gqlQueryMd5 = crypto.createHash('md5').update(JSON.stringify(redisValue)).digest('hex');
+
+    const redisQueryKey = `${REDIS_PREFIX}.${gqlQueryMd5}.body`;
+    const redisQueryResponse = `${REDIS_PREFIX}.${gqlQueryMd5}.response`;
+    const batchSize = Number(process.env.GUEST_CACHE_REDIS_BATCH_SIZE) || 3;
```
Comment (gabrielmatei) [inline packages/common/src/common/caching/guest-caching/guest-caching.middleware.ts:46]: i don't think is a good idea to read env variables from a library

File: packages/common/src/common/caching/guest-caching/guest-caching.service.ts (hunk 3 above)
Comment (florincoz) [reply packages/common/src/common/caching/guest-caching/guest-caching.service.ts:52]: Cannot use logger, it throws errors on large outputs or special characters

**Review by tanghel — APPROVED** [review]

**Pull request merged**